mindmeld/active_learning/heuristics.py

```diff
@@ -0,0 +1,490 @@
+        """
+        raise NotImplementedError("Subclasses must implement this method")
+
+
+class RandomSampling(ABC):
+    @staticmethod
+    def random_rank(num_elements: int) -> List[int]:
+        """Randomly shuffles indices.
+        Args:
+            num_elements (int): Number of elements to randomly sample.
+        Returns:
+            ranked_indices (List[int]): Indices corresponding to elements ranked by randomly.
+        """
+        ranked_indices = np.arange(num_elements)
+        np.random.shuffle(ranked_indices)
+        return list(ranked_indices)
+
+    @staticmethod
+    def rank_2d(confidences_2d: List[List[float]]) -> List[int]:
+        """Randomly shuffles indices.
+        Args:
+            confidences_2d (List[List[float]]): Confidence probabilities per element.
+        Returns:
+            ranked_indices (List[int]): Indices corresponding to elements ranked by utility.
+        """
+        num_elements = len(confidences_2d)
+        return RandomSampling.random_rank(num_elements)
+
+    @staticmethod
+    def rank_3d(confidences_3d: List[List[List[float]]]) -> List[int]:
+        """Randomly shuffles indices.
+        Args:
+            confidences_3d (List[List[List[float]]]): Confidence probabilities per element.
+        Returns:
+            ranked_indices (List[int]): Indices corresponding to elements ranked by utility.
+        """
+        _, num_elements, _ = np.array(confidences_3d).shape
+        return RandomSampling.random_rank(num_elements)
+
+
+class LeastConfidenceSampling(ABC):
+    @staticmethod
+    def rank_2d(confidences_2d: List[List[float]]) -> List[int]:
+        """First calculates the highest (max) confidences per element and then returns
+        the elements with the lowest max confidence.
+
+        Args:
+            confidences_2d (List[List[float]]): Confidence probabilities per element.
+        Returns:
+            ranked_indices (List[int]): Indices corresponding to elements ranked by utility.
+        """
+        highest_confidence_per_element = np.max(confidences_2d, axis=1)
+        return list(np.argsort(highest_confidence_per_element))
+
+    @staticmethod
+    def rank_3d(confidences_3d: List[List[List[float]]]) -> List[int]:
+        """First calculates the highest (max) confidences per element and then returns
+        the elements with the lowest max confidence. This is done for each confidence_2d
+        in a confidence_3d. The rankings are added to generate a final ranking.
+
+        Args:
+            confidences_3d (List[List[List[float]]]): Confidence probabilities per element.
+        Returns:
+            ranked_indices (List[int]): Indices corresponding to elements ranked by utility.
+        """
+        total_rank_score = np.sum(
+            [LeastConfidenceSampling.rank_2d(c) for c in confidences_3d], axis=0
+        )
+        high_to_low_ranked_indices = np.argsort(total_rank_score)[::-1]
```

Comment:
As discussed with @KKsharma99, `total_rank_score` gets classifier x samples ordered by least confidence. So we end up summing sample indices as opposed to sample ranks. Need to reshape the arrays before summing so we sum the sample ranks.

Reply:
Good catch! Will fix